pkg/master/controller/crdregistration/crdregistration_controller.go

```diff
@@ -58,17 +60,30 @@ type crdRegistrationController struct {
 	// queue is where incoming work is placed to de-dup and to allow "easy" rate limited requeues on errors
 	// this is actually keyed by a groupVersion
 	queue workqueue.RateLimitingInterface
+
+	// openAPIAggregationManager exposes interface to manager apiExtensionsServer's OpenAPI Services,
+	// which allows aggregating OpenAPI spec for new CRD API services with the existing static OpenAPI
+	// spec.
+	//
+	// TODO(roycaihw): NOTE we are re-using the AggregationManager used by aggregation apiserver to
+	// save some code, but openapi.AggregationManager is more complex than what we need: CRD openapi
+	// spec aggregation should have no conflict so we can simply add things up, instead of walking
+	// through all swagger  to resolve conflicts. However, currently we are not optimizing performance
+	// around CRDs because we don't have a lot of CRDs in cluster. We should keep in mind that this
```

Comment:
What does this mean if we have 100ish many CRDs?

Reply:
Every time one CRD gets added/updated/deleted, `openapi.AggregationManager` [builds the entire OpenAPI spec](https://github.com/kubernetes/kubernetes/blob/6d76e35b392a458c9c02e916326bba8e9c8ab8f4/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go#L188-L215) by 1. sorting the slice of spec by priority (not needed for CRDs because CRD apiservices are of same priority); 2. iterating through the specs and merging each spec (`source`) with the previous merged specs (`dest`). 

In each [merge process](https://github.com/kubernetes/kube-openapi/blob/e3762e86a74c878ffed47484592986685639c2cd/pkg/aggregator/aggregator.go#L256-L374) the aggregator iterates through `source.Paths` and `source.Definitions` a couple of times to check if there are any conflicts against `dest`, but it's also unnecessary for CRDs because CRD GVKs are unique. 

What we need can be something simple like: 
```
for k, v := range source.Definitions {
	dest.Definitions[k] = v
}
for k, v := range source.Paths.Paths {
	dest.Paths.Paths[k] = v
}
```

The main benefit of reusing `openapi.AggregationManager` is to save some code on recording the apiservices and specs. I will add another AggregationManager for CRDs

Reply:
"total 0.76s" sounds far too much. In any case, we should probably throttle this to once every few seconds. Wdyt?

Reply:
0.76s is the aggregation time for in-tree specs, which only happens at kube-apiserver start time. 

But I agree that we should have the simplified aggregator that doesn't check conflict. If we have 100ish CRDs and they get added to the cluster at start time, we will have a peak hitting the aggregation way more frequent than once every few seconds.